tools/rv_histogram.c

```diff
@@ -0,0 +1,276 @@
+
+typedef struct {
+    char insn_str[16];
+    size_t freq;
+} rv_insn_hist_t;
+
+static rv_insn_hist_t rv_insn_str_and_freq_tbl[] = {
+    {"nop", 0},       {"lui", 0},      {"auipc", 0},    {"jal", 0},
+    {"jalr", 0},      {"beq", 0},      {"bne", 0},      {"blt", 0},
+    {"bge", 0},       {"bltu", 0},     {"bgeu", 0},     {"lb", 0},
+    {"lh", 0},        {"lw", 0},       {"lbu", 0},      {"lhu", 0},
+    {"sb", 0},        {"sh", 0},       {"sw", 0},       {"addi", 0},
+    {"slti", 0},      {"sltiu", 0},    {"xori", 0},     {"ori", 0},
+    {"andi", 0},      {"slli", 0},     {"srli", 0},     {"srai", 0},
+    {"add", 0},       {"sub", 0},      {"sll", 0},      {"slt", 0},
+    {"sltu", 0},      {"xor", 0},      {"srl", 0},      {"sra", 0},
+    {"or", 0},        {"and", 0},      {"ecall", 0},    {"ebreak", 0},
+    {"wfi", 0},       {"uret", 0},     {"sret", 0},     {"hret", 0},
+    {"mret", 0},
+#if RV32_HAS(Zifencei)
+    {"fencei", 0},
+#endif
+#if RV32_HAS(Zicsr)
+    {"csrrw", 0},     {"csrrs", 0},    {"csrrc", 0},    {"csrrwi", 0},
+    {"csrrsi", 0},    {"csrrci", 0},
+#endif
+#if RV32_HAS(EXT_M)
+    {"mul", 0},       {"mulh", 0},     {"mulhsu", 0},   {"mulhu", 0},
+    {"div", 0},       {"divu", 0},     {"rem", 0},      {"remu", 0},
+#endif
+#if RV32_HAS(EXT_A)
+    {"lrw", 0},       {"scw", 0},      {"amoswapw", 0}, {"amoaddw", 0},
+    {"amoxorw", 0},   {"amoandw", 0},  {"amoorw", 0},   {"amominw", 0},
+    {"amomaxw", 0},   {"amominux", 0}, {"amomaxux", 0},
+#endif
+#if RV32_HAS(EXT_F)
+    {"flw", 0},       {"fsw", 0},      {"fmadds", 0},   {"fmsubs", 0},
+    {"fnmsubs", 0},   {"fnmadds", 0},  {"fadds", 0},    {"fsubs", 0},
+    {"fmuls", 0},     {"fdivs", 0},    {"fsqrts", 0},   {"fsgnjs", 0},
+    {"fsgnjns", 0},   {"fsgnjxs", 0},  {"fmins", 0},    {"fmaxs", 0},
+    {"fcvtws", 0},    {"fcvtwus", 0},  {"fmvxw", 0},    {"feqs", 0},
+    {"flts", 0},      {"fles", 0},     {"fclasss", 0},  {"fcvtsw", 0},
+    {"fcvtswu", 0},   {"fmvmx", 0},
+#endif
+#if RV32_HAS(EXT_C)
+    {"caddi4spn", 0}, {"clw", 0},      {"csw", 0},      {"cnop", 0},
+    {"caddi", 0},     {"cjal", 0},     {"cli", 0},      {"caddi16sp", 0},
+    {"clui", 0},      {"csrli", 0},    {"csrai", 0},    {"candi", 0},
+    {"csub", 0},      {"cxor", 0},     {"cor", 0},      {"cand", 0},
+    {"cj", 0},        {"cbeqz", 0},    {"cbnez", 0},    {"cslli", 0},
+    {"clwsp", 0},     {"cjr", 0},      {"cmv", 0},      {"cebreak", 0},
+    {"cjalr", 0},     {"cadd", 0},     {"cswsp", 0},
+#endif
+    {"unknown", 0}};
+
+static int cmp_dec(const void *a, const void *b)
+{
+    return ((rv_insn_hist_t *) b)->freq - ((rv_insn_hist_t *) a)->freq;
+}
+
+static int cmp_asc(const void *a, const void *b)
+{
+    return ((rv_insn_hist_t *) a)->freq - ((rv_insn_hist_t *) b)->freq;
+}
+
+/* used to adjust the length of histogram bar */
+static unsigned short get_win_max_col()
+{
+    struct winsize terminal_size;
+    ioctl(STDOUT_FILENO, TIOCGWINSZ, &terminal_size);
```

Comment:
This tool is supposed to be portable, and you shall consider the environments lacking of full POSIX compliance. e.g., MS-Windows.